subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java

```diff
@@ -60,51 +102,83 @@ public void run() {
-                            //received access request for lock that is already closed
-                            continue;
+                    ContendedAction contendedAction = contendedActions.get(lockId);
+                    if (contendedAction == null) {
+                        acceptConfirmationAsLockRequester(lockId, packet.getPort());
+                    } else {
+                        if (!contendedAction.running) {
+                            startLockReleaseAsLockHolder(contendedAction);
                         }
-                    } finally {
-                        lock.unlock();
+                        communicator.confirmUnlockRequest(packet);
                     }
-                    action.run();
+                    lock.unlock();
                 }
             }
         };
     }
 
+    private void startLockReleaseAsLockHolder(ContendedAction contendedAction) {
+        unlockActionExecutor.execute(contendedAction.action);
+        contendedAction.running = true;
+    }
+
+    private void acceptConfirmationAsLockRequester(long lockId, int port) {
+        unlocksConfirmedFrom.put(lockId, port);
+        LOGGER.debug("Gradle process at port {} confirmed unlock request for lock with id {}.", port, lockId);
+    }
+
     public void start(long lockId, Runnable whenContended) {
         lock.lock();
+        unlocksRequestedFrom.remove(lockId);
+        unlocksConfirmedFrom.remove(lockId);
         try {
             assertNotStopped();
             if (communicator == null) {
                 throw new IllegalStateException("Must initialize the handler by reserving the port first.");
             }
-            if (executor == null) {
-                executor = executorFactory.create("File lock request listener");
-                executor.execute(listener());
+            if (fileLockRequestListener == null) {
+                fileLockRequestListener = executorFactory.create("File lock request listener");
+                fileLockRequestListener.execute(listener());
+            }
+            if (unlockActionExecutor == null) {
+                unlockActionExecutor = executorFactory.create("File lock release action executor");
             }
             if (contendedActions.containsKey(lockId)) {
                 throw new UnsupportedOperationException("Multiple contention actions for a given lock are currently not supported.");
             }
-            contendedActions.put(lockId, whenContended);
+            contendedActions.put(lockId, new ContendedAction(whenContended));
         } finally {
             lock.unlock();
         }
     }
 
-    public void pingOwner(int port, long lockId, String displayName) {
-        getCommunicator().pingOwner(port, lockId, displayName);
+    public boolean maybePingOwner(int port, long lockId, String displayName, long timeElapsed) {
+        if (unlocksConfirmedFrom.containsKey(lockId) && unlocksConfirmedFrom.get(lockId) == port) {
```

Comment:
That throws a NPE if the entry does not exist, because of the comparison with an `int` it tries to unbox the null. I now simplified it like this:
`Integer.valueOf(port).equals(unlocksConfirmedFrom.get(lockId))`